src/main/java/com/hivemq/client/internal/mqtt/handler/subscribe/MqttSubscriptionHandler.java

```diff
@@ -313,10 +340,31 @@ private void readUnsubAck(final @NotNull ChannelHandlerContext ctx, final @NotNu
             }
         }
 
-        handleComplete(ctx, packetIdentifier);
+        completePending(ctx, statefulUnsubscribeWithFlow);
     }
 
-    private void handleComplete(final @NotNull ChannelHandlerContext ctx, final int packetIdentifier) {
+    private void completePending(
+            final @NotNull ChannelHandlerContext ctx, final @NotNull MqttSubOrUnsubWithFlow.Stateful oldPending) {
+
+        if (firstPending == oldPending) {
+            firstPending = oldPending.next;
+            if (lastPending == oldPending) {
+                lastPending = null;
+            }
+        } else {
+            MqttSubOrUnsubWithFlow.Stateful beforePending = firstPending;
+            assert beforePending != null;
+            while (beforePending.next != oldPending) {
```

Comment:
When the client sends 2 subscribes and the server does not acknowledge them in the same order, the linked list (kind of) has to be searched.
As the oldPending has to be in this list (ensured if completePending is called) it can not throw a NPE.

Reply:
In the OutgoingQosHandler this is solved by a double linked list, probably this should be done here too, as it is a lot easier to understand

Reply:
Ah ok got it, yeah a oldPending.previous would be more readable

Reply:
replaced with double links (next, prev) now